File: server/routes/gis.ts
import { Router } from 'express';
import { prisma } from '../utils/prisma.js';
import { authGuard, requireRole } from '../middleware/auth.js';
import { writeAudit } from '../middleware/audit.js';

const router = Router();

// GET /diagram/:diagramId — get all GIS data for a diagram
router.get('/diagram/:diagramId', authGuard, async (req, res) => {
  const { diagramId } = req.params;

  const gisData = await prisma.gisData.findMany({
    where: {
      diagramInstance: { diagramId },
    },
    include: {
      diagramInstance: {
        select: { id: true, label: true, componentId: true },
      },
    },
  });

  res.json({ items: gisData });
});

// GET /:id — get single GIS data
router.get('/:id', authGuard, async (req, res) => {
  const { id } = req.params;

  const gis = await prisma.gisData.findUnique({
    where: { id },
    include: {
      diagramInstance: {
        select: { id: true, label: true, componentId: true, diagramId: true },
      },
    },
  });

  if (!gis) {
    res.status(404).json({ message: 'GIS 数据不存在' });
    return;
  }

  res.json(gis);
});

// PUT /instance/:instanceId — upsert GIS data by instance ID
router.put('/instance/:instanceId', authGuard, requireRole('ADMIN', 'DIAGRAM_EDITOR', 'GIS_EDITOR'), async (req, res) => {
  const { instanceId } = req.params;
  const { latitude, longitude } = req.body ?? {};

  // Verify instance exists
  const instance = await prisma.diagramInstance.findUnique({ where: { id: instanceId } });
  if (!instance) {
    res.status(404).json({ message: '实例不存在' });
    return;
  }

  const data = {
    latitude: latitude !== undefined ? Number(latitude) : null,
    longitude: longitude !== undefined ? Number(longitude) : null,
    updatedBy: req.user!.id,
  };

  const gis = await prisma.gisData.upsert({
    where: { diagramInstanceId: instanceId },
    update: data,
    create: {
      diagramInstanceId: instanceId,
      ...data,
    },
  });

  await writeAudit(req.user!.id, 'GIS_DATA_UPSERT', 'GisData', gis.id, {
    instanceId,
    latitude: data.latitude,
    longitude: data.longitude,
  });

  res.json(gis);
});

// POST /batch — batch upsert GIS data
router.post('/batch', authGuard, requireRole('ADMIN', 'DIAGRAM_EDITOR', 'GIS_EDITOR'), async (req, res, next) => {
  try {
    const { items } = req.body as { items: Array<{
      diagramInstanceId: string;
      latitude?: number | null;
      longitude?: number | null;
    }> };
    if (!Array.isArray(items) || items.length === 0) {
      res.status(400).json({ message: 'items 不能为空数组' });
      return;
    }
    if (items.length > 500) {
      res.status(400).json({ message: '单次批量操作不能超过 500 条' });
      return;
    }

    const results = await Promise.all(
      items.map((item) =>
        prisma.gisData.upsert({
          where: { diagramInstanceId: item.diagramInstanceId },
          create: {
            diagramInstanceId: item.diagramInstanceId,
            latitude: item.latitude ?? null,
            longitude: item.longitude ?? null,
            updatedBy: req.user!.id,
          },
          update: {
            ...(item.latitude !== undefined ? { latitude: item.latitude } : {}),
            ...(item.longitude !== undefined ? { longitude: item.longitude } : {}),
            updatedBy: req.user!.id,
          },
        })
      )
    );
    res.json({ count: results.length });
  } catch (err) { next(err); }
});

export default router;

File: src-tauri/src/commands/gis.rs
use ecdraw_core::error::AppError;
use ecdraw_core::middleware;
use ecdraw_core::models::GisData;
use ecdraw_core::AppState;
use serde::{Deserialize, Serialize};
use serde_json::json;
use tauri::State;
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BatchGisItem {
    pub diagram_instance_id: String,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
}

/// GET /api/gis/diagram/:diagramId
#[tauri::command]
pub async fn list_gis_by_diagram(
    state: State<'_, AppState>,
    token: String,
    diagram_id: String,
) -> Result<Vec<serde_json::Value>, AppError> {
    let _claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    let did: Uuid = diagram_id.parse().map_err(|_| AppError::BadRequest("无效的图纸ID".into()))?;

    let items = sqlx::query_as::<_, GisData>(
        r#"SELECT g.* FROM gis_data g
           JOIN diagram_instances di ON g.diagram_instance_id = di.id
           WHERE di.diagram_id = $1"#
    )
    .bind(did)
    .fetch_all(&state.pool)
    .await?;

    let results: Vec<serde_json::Value> = items.iter().map(|data| {
        json!({
            "id": data.id,
            "diagramInstanceId": data.diagram_instance_id,
            "latitude": data.latitude,
            "longitude": data.longitude,
        })
    }).collect();

    Ok(results)
}

/// PUT /api/gis/instance/:instanceId
#[tauri::command]
pub async fn upsert_gis(
    state: State<'_, AppState>,
    token: String,
    instance_id: String,
    latitude: Option<f64>,
    longitude: Option<f64>,
) -> Result<GisData, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    middleware::require_role(&claims, &["ADMIN", "DIAGRAM_EDITOR", "GIS_EDITOR"])?;
    let user_id: Uuid = claims.sub.parse().unwrap();
    let iid: Uuid = instance_id.parse().map_err(|_| AppError::BadRequest("无效的实例ID".into()))?;

    let _inst = sqlx::query_scalar::<_, Uuid>("SELECT id FROM diagram_instances WHERE id = $1")
        .bind(iid).fetch_optional(&state.pool).await?
        .ok_or_else(|| AppError::NotFound("实例不存在".into()))?;

    let data = sqlx::query_as::<_, GisData>(
        r#"INSERT INTO gis_data (diagram_instance_id, latitude, longitude, updated_by)
           VALUES ($1, $2, $3, $4)
           ON CONFLICT (diagram_instance_id)
           DO UPDATE SET latitude = $2, longitude = $3, updated_by = $4, updated_at = NOW()
           RETURNING *"#
    )
    .bind(iid).bind(latitude).bind(longitude).bind(user_id)
    .fetch_one(&state.pool)
    .await?;

    Ok(data)
}

/// POST /api/gis/batch
#[tauri::command]
pub async fn batch_upsert_gis(
    state: State<'_, AppState>,
    token: String,
    items: Vec<BatchGisItem>,
) -> Result<i32, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    middleware::require_role(&claims, &["ADMIN", "DIAGRAM_EDITOR", "GIS_EDITOR"])?;
    let user_id: Uuid = claims.sub.parse().unwrap();

    if items.len() > 500 {
        return Err(AppError::BadRequest("单次最多导入500条".into()));
    }

    let mut count = 0;
    for item in &items {
        let iid: Uuid = item.diagram_instance_id.parse().map_err(|_| AppError::BadRequest("无效的实例ID".into()))?;
        let r = sqlx::query(
            r#"INSERT INTO gis_data (diagram_instance_id, latitude, longitude, updated_by)
               VALUES ($1, $2, $3, $4)
               ON CONFLICT (diagram_instance_id)
               DO UPDATE SET latitude = $2, longitude = $3, updated_by = $4, updated_at = NOW()"#
        )
        .bind(iid).bind(item.latitude).bind(item.longitude).bind(user_id)
        .execute(&state.pool).await?;
        count += r.rows_affected() as i32;
    }

    Ok(count)
}
